# Importing libraries 

In [1]:
#Importing libraries 
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import os

In [2]:
# Data path
path =r'C:\Users\ANITA BOADU\New York City Bikes'

In [3]:
df=pd.read_csv(os.path.join(path,'Data','Original Data','citibike.csv'),index_col=False)

In [4]:
print(df.columns)

Index(['trip_id', 'bike_id', 'weekday', 'start_hour', 'start_time',
       'start_station_id', 'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_time', 'end_station_id',
       'end_station_name', 'end_station_latitude', 'end_station_longitude',
       'trip_duration', 'subscriber', 'birth_year', 'gender'],
      dtype='object')


In [5]:
# Basic info and summary
print(df.info())
print(df.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   trip_id                  50000 non-null  object 
 1   bike_id                  50000 non-null  int64  
 2   weekday                  50000 non-null  object 
 3   start_hour               50000 non-null  int64  
 4   start_time               50000 non-null  object 
 5   start_station_id         50000 non-null  int64  
 6   start_station_name       50000 non-null  object 
 7   start_station_latitude   50000 non-null  float64
 8   start_station_longitude  50000 non-null  float64
 9   end_time                 50000 non-null  object 
 10  end_station_id           50000 non-null  int64  
 11  end_station_name         50000 non-null  object 
 12  end_station_latitude     50000 non-null  float64
 13  end_station_longitude    50000 non-null  float64
 14  trip_duration         

In [6]:
# Cell 4: Convert times to datetime
df['start_time'] = pd.to_datetime(df['start_time'], dayfirst=True)
df['end_time'] = pd.to_datetime(df['end_time'], dayfirst=True)


# Calculate trip duration in minutes (already given but ensure correctness)
df['trip_duration_min'] = df['trip_duration'] / 60

In [7]:
# Cell 5: Feature engineering
# Extract hour and weekday if not already numerical
df['start_hour'] = df['start_time'].dt.hour
df['weekday'] = df['start_time'].dt.day_name()

In [8]:
# Cell 6: Check missing values
df.isnull().sum()

trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
trip_duration_min             0
dtype: int64

In [9]:
df.head()

trip_id  bike_id    weekday  start_hour          start_time  \
0  LnQzQk    16013     Monday          18 2013-09-09 18:18:00   
1  IL9boN    15230   Thursday          18 2013-09-12 18:38:00   
2  46clGB    17942  Wednesday          19 2013-09-18 19:44:00   
3  v7vdFt    19683   Saturday          11 2013-09-28 11:54:00   
4  VGBsb5    18024   Saturday          18 2013-09-07 18:08:00   

   start_station_id       start_station_name  start_station_latitude  \
0               523          W 38 St & 8 Ave               40.754666   
1               257  Lispenard St & Broadway               40.719392   
2               479          9 Ave & W 45 St               40.760193   
3               527          E 33 St & 1 Ave               40.743156   
4               521          8 Ave & W 31 St               40.750450   

   start_station_longitude            end_time  end_station_id  \
0               -73.991382 2013-09-09 18:35:00             334   
1               -74.002472 2013-09-12 18:48:00             236   
2               -73.991255 2013-09-18 19:50:00             513   
3               -73.974347 2013-09-28 12:03:00             441   
4               -73.994811 2013-09-07 18:46:00             476   

      end_station_name  end_station_latitude  end_station_longitude  \
0      W 20 St & 7 Ave             40.742388             -73.997262   
1  St Marks Pl & 2 Ave             40.728419             -73.987140   
2     W 56 St & 10 Ave             40.768254             -73.988639   
3      E 52 St & 2 Ave             40.756014             -73.967416   
4      E 31 St & 3 Ave             40.743943             -73.979661   

   trip_duration      subscriber  birth_year  gender  trip_duration_min  
0            993      Subscriber      1968.0       2          16.550000  
1            581      Subscriber      1983.0       1           9.683333  
2            361      Subscriber      1989.0       1           6.016667  
3            561      Subscriber      1988.0       2           9.350000  
4           2296  Non-Subscriber         NaN       0          38.266667

In [10]:
# Check data types (especially datetime columns
df.dtypes

trip_id                            object
bike_id                             int64
weekday                            object
start_hour                          int32
start_time                 datetime64[ns]
start_station_id                    int64
start_station_name                 object
start_station_latitude            float64
start_station_longitude           float64
end_time                   datetime64[ns]
end_station_id                      int64
end_station_name                   object
end_station_latitude              float64
end_station_longitude             float64
trip_duration                       int64
subscriber                         object
birth_year                        float64
gender                              int64
trip_duration_min                 float64
dtype: object

In [11]:
#Convert datetime columns:

df['start_time'] = pd.to_datetime(df['start_time'])
df['end_time'] = pd.to_datetime(df['end_time'])

In [12]:
# Optional: calculate age from birth year (handling missing values):

from datetime import datetime
current_year = datetime.now().year
df['age'] = current_year - df['birth_year']


In [13]:
# Quick stats for numerical columns:

df.describe()

bike_id    start_hour                  start_time  \
count  50000.000000  50000.000000                       50000   
mean   17615.269360     14.145240  2013-09-16 14:31:04.322400   
min    14556.000000      0.000000         2013-09-01 00:00:00   
25%    16188.000000     10.000000         2013-09-09 10:25:00   
50%    17584.000000     15.000000         2013-09-16 18:54:00   
75%    19014.000000     18.000000         2013-09-23 21:55:15   
max    20642.000000     23.000000         2013-09-30 23:58:00   
std     1675.407446      4.860541                         NaN   

       start_station_id  start_station_latitude  start_station_longitude  \
count      50000.000000            50000.000000             50000.000000   
mean         443.321500               40.734170               -73.991109   
min           72.000000               40.680342               -74.017134   
25%          304.000000               40.720196               -74.000271   
50%          402.000000               40.735877               -73.990765   
75%          484.000000               40.750020               -73.981923   
max         3002.000000               40.770513               -73.950048   
std          356.559925                0.019911                 0.012555   

                            end_time  end_station_id  end_station_latitude  \
count                          50000    50000.000000          50000.000000   
mean   2013-09-16 14:46:16.427999744      442.539700             40.733859   
min              2013-09-01 00:07:00       72.000000             40.680342   
25%              2013-09-09 10:40:30      304.000000             40.720196   
50%              2013-09-16 19:11:00      402.000000             40.735354   
75%              2013-09-23 22:06:15      483.000000             40.749013   
max              2013-10-01 00:15:00     3002.000000             40.770513   
std                              NaN      355.756022              0.019885   

       end_station_longitude  trip_duration    birth_year        gender  \
count           50000.000000   50000.000000  43021.000000  50000.000000   
mean              -73.991351     838.982900   1975.627786      1.073540   
min               -74.017134      60.000000   1899.000000      0.000000   
25%               -74.001547     417.000000   1968.000000      1.000000   
50%               -73.991218     672.000000   1978.000000      1.000000   
75%               -73.982050    1112.000000   1984.000000      1.000000   
max               -73.950048    2697.000000   1997.000000      2.000000   
std                 0.012569     573.663997     11.089001      0.589389   

       trip_duration_min           age  
count       50000.000000  43021.000000  
mean           13.983048     50.372214  
min             1.000000     29.000000  
25%             6.950000     42.000000  
50%            11.200000     48.000000  
75%            18.533333     58.000000  
max            44.950000    127.000000  
std             9.561067     11.089001

##### Observation from df.describe ()
#######Trip IDs and Bike IDs
-bike_id ranges from 14556 to 20642 with a mean of 17615.
-Indicates a large number of unique bikes in the dataset.

####### Start Hour 
-Values range from 0 to 23, representing all hours of the day.
-Mean is around 14, so most trips happen in the early afternoon.

####### Datetime Columns (start_time, end_time)
-Columns are objects, but min/max values show trips occurred between
2013-09-01 and 2013-10-01.
-Consider converting these columns to datetime for better analysis

######## Stations
-start_station_id ranges from 72 to 3002, showing many stations.
-Latitude and longitude values confirm all stations are in NYC bounds
(~40.68–40.77, -74.02–-73.95).

#######Trip Duration 
-trip_duration ranges from 60 to 2697 seconds (1 to ~45 minutes), mean ~**839 seconds (14 minutes)**.
-trip_duration_min column confirms similar stats in minutes.

#######User Info
-birth_year has missing values (43,021 non-null out of 50,000).
-Birth years range 1899–1997, mean ~1976 → typical adult riders.
-age column is pre-calculated; mean ~49, max 126 (possible outliers).

####### Gender
-Encoded as 0, 1, 2. Mean ~1.07, likely indicating more male users (if 1 = male).



####### Data Summary 

-Dataset contains 50,000 trips with information on bike IDs, start/end stations, times, user demographics, and trip duration.
-Most trips occur in the afternoon (~2 PM).
-Trip durations vary widely, from 1 minute to 45 minutes.
-Some birth year values are missing (around 7,000 rows), and age shows outliers (max 126), which may require cleaning.
-Start/end station coordinates confirm all trips occur within NYC bounds.

In [14]:
# Check if there are duplicate rows
duplicates = df[df.duplicated()]

# Show how many and preview them
print(f"Number of duplicate rows: {duplicates.shape[0]}")
duplicates.head()

Number of duplicate rows: 0


Empty DataFrame
Columns: [trip_id, bike_id, weekday, start_hour, start_time, start_station_id, start_station_name, start_station_latitude, start_station_longitude, end_time, end_station_id, end_station_name, end_station_latitude, end_station_longitude, trip_duration, subscriber, birth_year, gender, trip_duration_min, age]
Index: []

In [15]:
# Check missing values in each column
missing_values = df.isnull().sum()

# Display columns with missing values
print("Missing values per column:\n", missing_values)

Missing values per column:
 trip_id                       0
bike_id                       0
weekday                       0
start_hour                    0
start_time                    0
start_station_id              0
start_station_name            0
start_station_latitude        0
start_station_longitude       0
end_time                      0
end_station_id                0
end_station_name              0
end_station_latitude          0
end_station_longitude         0
trip_duration                 0
subscriber                    0
birth_year                 6979
gender                        0
trip_duration_min             0
age                        6979
dtype: int64


In [16]:
# 1️⃣ Calculate age
current_year = datetime.now().year
df['age'] = current_year - df['birth_year']

# making year again into integer

df['birth_year'] = df['birth_year'].astype('Int64')


# 2️⃣ Fill missing ages with median age (convert to int here)
median_age = int(df['age'].median())  # ensure median is an int
df['age'] = df['age'].fillna(median_age).astype(int)

# 3️⃣ Verify results
print("Missing values after filling age:")
print(df[['birth_year', 'age']].isnull().sum())

# Optional: preview
df[['birth_year', 'age']].head()

Missing values after filling age:
birth_year    6979
age              0
dtype: int64


birth_year  age
0        1968   58
1        1983   43
2        1989   37
3        1988   38
4        <NA>   48

### Note on Feature Engineering and Missing Values

####### Age Calculation

Original column: birth_year.
Calculated age as: age = current_year - birth_year.
Why:
Makes user demographics meaningful.
Useful for research questions like:
Are younger or older people taking more trips?
How does trip duration vary with age?
Easier for plotting distributions, correlations, and segmenting users.
Handling missing birth_year:
Missing values in age are filled with the median age.
age now has no missing values, even though birth_year still has some missing entries.

Trip Duration in Minutes

Original column: trip_duration (in seconds).
Calculated trip_duration_min = trip_duration / 60.
Why:
Makes trip lengths human-readable (e.g., 14 minutes instead of 839 seconds).
Easier for plotting distributions and comparing trips.
Helps answer questions like:
Most common trip duration.
Differences between weekdays and weekends.

Handling Missing birth_year

birth_year still has missing values (~6,979 rows).
Decision:
No need to fill birth_year since we already have age for analysis.
Use age for all demographic and trip-duration analyses.
Only fill or drop birth_year if exact birth years are specifically needed.

In [17]:
import os

# 1. Define your base project path
path = r'C:\Users\ANITA BOADU\New York City Bikes'

# 2. Define the subfolder where you want to save the clean data
# It is standard practice to separate 'Original Data' from 'Prepared Data'
export_folder = os.path.join(path, 'Data', 'Prepared Data')

# 3. Create the 'Prepared Data' folder if it doesn't exist
if not os.path.exists(export_folder):
    os.makedirs(export_folder)
    print(f"Created folder: {export_folder}")

# 4. Define the final file name and save
file_name = 'citibike_cleaned.csv'
df.to_csv(os.path.join(export_folder, file_name), index=False)

print(f"Success! Your cleaned data is saved here: {os.path.join(export_folder, file_name)}")

Success! Your cleaned data is saved here: C:\Users\ANITA BOADU\New York City Bikes\Data\Prepared Data\citibike_cleaned.csv
